File: Rare6_이분탐색/juny/9007_카누선수.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>
using namespace std;

#define endl '\n'

int T, k, n;

int main() {
    ios_base :: sync_with_stdio(false); cin.tie(NULL);

    cin >> T;
    while(T--) {
        cin >> k >> n;
        vector< vector<int> > v(4, vector<int>(n));

        for(int i = 0; i < 4; i++) {
            for(int j = 0; j < n; j++) cin >> v[i][j];
        }
        vector<int> list1(n*n), list2(n*n);
        int idx = 0;
        for(int i = 0; i < n; i++) {
            for(int j = 0; j < n; j++) {
                list1[idx] = v[0][i] + v[1][j];
                list2[idx] = v[2][i] + v[3][j];
                idx++;
            }
        }

        sort(list1.begin(), list1.end());
        sort(list2.begin(), list2.end());

        int min = 987654321, res = 0;
        bool flag = false;
        for(int i = 0; i < n*n; i++) {
            int k2 = k - list1[i];
            int left = 0, right = (n*n)-1;
            while(left <= right) {
                int mid = (left + right) / 2;
                if(min == abs(k2 - list2[mid]) && res > list1[i] + list2[mid])
                    res = list1[i] + list2[mid];
                else if(min > abs(k2 - list2[mid])) {
                    min = abs(k2 - list2[mid]);
                    res = list1[i] + list2[mid];
                }

                if(k2 < list2[mid]) right = mid - 1;
                else if(k2 > list2[mid]) left = mid + 1;
                else {
                    cout << k << endl;
                    flag = true;
                    break;
                }
            }
            if(flag) break;
        }
        if(!flag) cout << res << endl;
    }

    return 0;
}

File: Rare6_이분탐색/wons/9007_카누선수.py
import sys

read = sys.stdin.readline

T = int(read())
for _ in range(T) :
    K,N = map(int,read().split())

    stu = [list(map(int,read().split())) for _ in range(4)]
    c1,c2 = [],[]
    for i in range(N) :
        for j in range(N) :
            c1.append(stu[0][i] + stu[1][j])
            c2.append(stu[2][i] + stu[3][j])

    c1.sort(); c2.sort()

    cmp = int(1e9)
    res = 0
    for c in c1 :
        weight = K - c

        left = 0; right = len(c1)-1
        while left <= right :
            mid = (left+right)//2

            if cmp == abs(weight - c2[mid]) and res > c + c2[mid]:
                res = c+c2[mid]
            
            elif cmp > abs(weight - c2[mid]) :
                cmp = abs(weight - c2[mid])
                res = c + c2[mid]
            
            if weight > c2[mid] :
                left = mid + 1
            else :
                right = mid - 1
    print(res)

    



